src/optimizer/cardinality_estimator.cpp

```diff
@@ -0,0 +1,477 @@
+                                         vector<column_binding_map_t<ColumnBinding>> &child_binding_maps) {
+	for (auto &map_set : child_binding_maps) {
+		for (auto &mapping : map_set) {
+			ColumnBinding relation_bindings = mapping.first;
+			ColumnBinding actual_bindings = mapping.second;
+
+			if (actual_bindings.table_index == binding_index) {
+				auto key = ColumnBinding(relation_id, relation_bindings.column_index);
+				AddRelationToColumnMapping(key, actual_bindings);
+			}
+		}
+	}
+}
+
+void CardinalityEstimator::InitCardinalityEstimatorProps(vector<struct NodeOp> *node_ops,
+                                                         vector<unique_ptr<FilterInfo>> *filter_infos) {
+	InitEquivalentRelations(filter_infos);
+	InitTotalDomains();
+	for (idx_t i = 0; i < node_ops->size(); i++) {
+		auto join_node = (*node_ops)[i].node.get();
+		auto op = (*node_ops)[i].op;
+		join_node->SetBaseTableCardinality(op->EstimateCardinality(context));
+		if (op->type == LogicalOperatorType::LOGICAL_COMPARISON_JOIN) {
+			auto &join = (LogicalComparisonJoin &)*op;
+			if (join.join_type == JoinType::LEFT) {
+				// TODO: inspect child operators to get a more accurate cost
+				// and cardinality estimation. If an base op is a Logical Comparison join
+				// it is probably a left join, so cost of the larger table is a fine
+				// estimate
+				// No need to update a mark join cost because I say so.
+				join_node->SetCost(join_node->GetBaseTableCardinality());
+			}
+		}
+		// update cardinality with filters
+		EstimateBaseTableCardinality(join_node, op);
+		UpdateTotalDomains(join_node, op);
+	}
+
+	AssertEquivalentRelationSize();
+}
+
+void CardinalityEstimator::UpdateTotalDomains(JoinNode *node, LogicalOperator *op) {
+	auto relation_id = node->set->relations[0];
+	TableCatalogEntry *catalog_table = nullptr;
+	auto get = GetLogicalGet(op);
+	if (get) {
+		catalog_table = GetCatalogTableEntry(get);
+	}
+
+	//! Initialize the tdoms for all columns the relation uses in join conditions.
+	unordered_set<idx_t>::iterator ite;
+	idx_t count = node->GetBaseTableCardinality();
+
+	bool direct_filter = false;
+	for (auto &column : relation_attributes[relation_id].columns) {
+		//! for every column in the relation, get the count via either HLL, or assume it to be
+		//! the cardinality
+		ColumnBinding key = ColumnBinding(relation_id, column);
+
+		// TODO: Go through table filters and find if there is a direct filter
+		//  on a join filter
+		if (catalog_table) {
+			relation_attributes[relation_id].original_name = catalog_table->name;
+			// Get HLL stats here
+			auto actual_binding = relation_column_to_original_column[key];
+
+			// sometimes base stats is null (test_709.test) returns null for base stats while
+			// there is still a catalog table. Anybody know anything about this?
+			auto base_stats = catalog_table->storage->GetStatistics(context, actual_binding.column_index);
```

Comment:
We can also call the `statistics` method in the `TableFunction` in the `LogicalGet` - that way this works for other scans besides only the base table scans. We might consider emitting distinct count statistics for Parquet files as well, for example.